File: src/commands/check.ts
/**
 * `hal check` — Stop hook gate.
 *
 * Exit codes:
 *   0 = allow (agent can stop)
 *   2 = block (agent must continue)
 */

import { readFileSync } from "node:fs";
import { HINTS } from "../lib/hints.js";
import { buildResponse, formatResponse } from "../lib/response-builder.js";
import { getTaskPathForSession } from "../lib/session-manager.js";
import { loadAndValidateTask } from "../lib/validator.js";

/**
 * Handle the check command.
 * Reads session_id from stdin JSON.
 */
export function handleCheck(): void {
	let stdinContent: string;
	try {
		stdinContent = readFileSync(0, "utf-8").trim();
	} catch {
		process.exit(0);
	}

	if (!stdinContent) {
		process.exit(0);
	}

	let stdinData: { session_id?: string };
	try {
		stdinData = JSON.parse(stdinContent);
	} catch {
		process.exit(0);
	}

	const sessionId = stdinData.session_id;
	if (!sessionId) {
		process.exit(0);
	}

	const taskPath = getTaskPathForSession(sessionId);
	if (!taskPath) {
		process.exit(0);
	}

	let task: import("../types.js").TaskYaml;
	try {
		task = loadAndValidateTask(taskPath);
	} catch {
		process.exit(0);
	}

	// Check if paused (copilot mode) — allow stop
	const lastHistoryEvent =
		task.history && task.history.length > 0
			? task.history[task.history.length - 1]
			: null;
	if (lastHistoryEvent?.type === "paused") {
		process.exit(0);
	}

	// Check if all steps are done — allow stop
	const steps = task.steps ?? [];
	const allDone =
		steps.length > 0 && steps.every((s) => s.status === "done");
	if (allDone) {
		process.exit(0);
	}

	// Task status is pending or done — allow stop
	if (task.status === "pending" || task.status === "done") {
		process.exit(0);
	}

	// Block — remaining steps
	const remainingSteps = steps.filter((s) => s.status !== "done");

	const response = buildResponse({
		status: "blocked",
		message: "Incomplete steps remain",
		data: {
			task_goal: task.goal,
			task_status: task.status ?? "pending",
			remaining_steps: remainingSteps.map((s) => ({
				id: s.id,
				goal: s.goal,
				status: s.status ?? "pending",
			})),
		},
		commands_hint: HINTS.CHECK_BLOCKED,
	});

	console.error(formatResponse(response));
	process.exit(2);
}

File: src/commands/check.rs
use std::collections::BTreeMap;
use std::io::Read;

use crate::core::hints;
use crate::core::response::HalResponse;
use crate::core::session::get_task_path_for_session;
use crate::core::validator::load_and_validate_task;
use crate::types::Status;

pub fn handle_check() {
    let mut stdin_content = String::new();
    if std::io::stdin().read_to_string(&mut stdin_content).is_err() || stdin_content.trim().is_empty() {
        std::process::exit(0);
    }

    let stdin_data: serde_json::Value = match serde_json::from_str(stdin_content.trim()) {
        Ok(v) => v,
        Err(_) => std::process::exit(0),
    };

    let session_id = match stdin_data.get("session_id").and_then(|v| v.as_str()) {
        Some(id) => id.to_string(),
        None => std::process::exit(0),
    };

    let task_path = match get_task_path_for_session(&session_id) {
        Some(p) => p,
        None => std::process::exit(0),
    };

    let task = match load_and_validate_task(std::path::Path::new(&task_path)) {
        Ok(t) => t,
        Err(_) => std::process::exit(0),
    };

    // Paused -> allow
    if let Some(history) = &task.history {
        if let Some(last) = history.last() {
            if matches!(last, crate::types::HistoryEvent::Paused { .. }) {
                std::process::exit(0);
            }
        }
    }

    let steps = task.steps.as_deref().unwrap_or(&[]);
    let all_done = !steps.is_empty() && steps.iter().all(|s| s.status.as_ref() == Some(&Status::Done));
    if all_done {
        std::process::exit(0);
    }

    let task_status = task.status.as_ref().unwrap_or(&Status::Pending);
    if *task_status == Status::Pending || *task_status == Status::Done {
        std::process::exit(0);
    }

    // Block
    let remaining: Vec<serde_yaml::Value> = steps.iter()
        .filter(|s| s.status.as_ref() != Some(&Status::Done))
        .map(|s| {
            let mut m = serde_yaml::Mapping::new();
            m.insert("id".into(), serde_yaml::Value::String(s.id.clone()));
            m.insert("goal".into(), serde_yaml::Value::String(s.goal.clone()));
            m.insert("status".into(), serde_yaml::Value::String(s.status.as_ref().unwrap_or(&Status::Pending).to_string()));
            serde_yaml::Value::Mapping(m)
        })
        .collect();

    let mut data = BTreeMap::new();
    data.insert("task_goal".into(), serde_yaml::Value::String(task.goal));
    data.insert("task_status".into(), serde_yaml::Value::String(task_status.to_string()));
    data.insert("remaining_steps".into(), serde_yaml::Value::Sequence(remaining));

    let response = HalResponse::new("blocked", "Incomplete steps remain")
        .with_data(data)
        .with_hint(hints::CHECK_BLOCKED);

    eprintln!("{}", response.to_yaml());
    std::process::exit(2);
}
